src/reducers/computeRenderedColumns.js

```diff
@@ -0,0 +1,244 @@
+/**
+ * Copyright Schrodinger, LLC
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule computeRenderedColumns
+ */
+
+'use strict';
+
+import clamp from 'lodash/clamp';
+import columnWidths from 'columnWidths';
+import scrollbarsVisibleSelector from 'scrollbarsVisible';
+import updateColumnWidth from 'updateColumnWidth';
+
+export default function computeRenderedColumns(state, columnAnchor) {
+  // clone state
+  const newState = Object.assign({}, state);
+  
+  // get buffer and viewport range
+  const columnRange = calculateRenderedColumnRange(newState, columnAnchor);
+  
+  // update the offsets and buffer mapping
+  computeRenderedColumnOffsets(newState, columnRange, newState.scrolling);
+
+  // scrollX might have changed due to change in columns and offsets
+  const { scrollableColumns, maxScrollX } = columnWidths(state);
+  let scrollX = 0;
+  if (scrollableColumns.length > 0) {
+    scrollX = newState.columnOffsets[columnRange.firstViewportIdy] - newState.firstColumnOffset;
+  }
+  newState.scrollX = clamp(scrollX, 0, maxScrollX);
+
+  return newState;
+}
+
+/**
+ * Determine the range of columns to render (buffer and viewport)
+ * The leading and trailing buffer is based on a fixed count,
+ * while the viewport columns are based on their height and the viewport height
+ * We use the columnAnchor to determine what either the first or last column
+ * will be, as well as the offset.
+ *
+ * NOTE (jordan) This alters state so it shouldn't be called
+ * without state having been cloned first.
+ *
+ * @param {!Object} state
+ * @param {{
+ *   firstIndex: number,
+ *   firstOffset: number,
+ *   lastIndex: number,
+ * }} columnAnchor
+ * @return {{
+ *   endBufferIdy: number,
+ *   endViewportIdy: number,
+ *   firstBufferIdy: number,
+ *   firstViewportIdy: number,
+ * }}
+ * @private
+ */
+function calculateRenderedColumnRange(state, columnAnchor) {
+  const bufferColumnCount = 3; // TODO (pradeep): should we calculate this similar to bufferColumnCount ?
+  const { availableScrollWidth, scrollableColumns } = columnWidths(state);
+  const columnCount = scrollableColumns.length;
+
+  if (columnCount === 0) {
+    return {
+      endBufferIdy: 0,
```

Comment:
should this be Idx rather than Idy, here and throughout?

Reply:
I'm not a fan of that since Y never appears in the work index which Idx is short for...

If we do use why, let's do colY instead of idY

Reply:
> I'm not a fan of that since Y never appears in the work index which Idx is short for...

Ah wow I completely missed that idx is short for index...

> If we do use why, let's do colY instead of idY

`endBufferColY` sounds good to me.

Since we have col now, we could skip Y and just put `endBufferColumn` or `endBufferCol`
or we can still have idx so that we get `endViewportColumnIdx`, but it seems a bit too long...

So, which one do you prefer?
* `endBufferColY`
* `endBufferCol`
* `endBufferColumn`
* `endViewportColumnIdx`

Reply:
endBufferCol sounds good to me.  But I'm happy with all of those options.